File: src/building/xml/Skill.ts
import { parseIntOrDefault } from "../util/strings.js";
import { isTransfiguredSkill } from "../transform/gem.js";
import * as itemTypes from "../../api/item.js";

export class Skills {
    skillSet = new SkillSet();

    public toString(): string {
        return `<Skills activeSkillSet="1">
${this.skillSet}
</Skills>`;
    }
}

export class SkillSet {
    skills: Skill[] = [];
    public toString(): string {
        const skillsView = this.skills
            .map((skill) => skill.toString())
            .join("\n");
        return `<SkillSet id="1">
${skillsView}
</SkillSet>`;
    }
}

// 1. skills only depend on link, two active gems can be in one skill
// 2. vaal gem is treated as one gem
// 3. computed gem(like Arcanist Brand) is treated as one gem
export class Skill {
    slot = "";
    gems: Gem[] = [];
    imbuedSupport = "";

    constructor(slotName: string, jsonList: itemTypes.Gem[]) {
        this.slot = slotName;
        for (const json of jsonList) {
            const { gem, buildInSupport } = Gem.parseJson(json);
            this.gems.push(gem);
            if (buildInSupport) {
                this.imbuedSupport = buildInSupport.nameSpec;
            }
        }
    }

    public toString(): string {
        const gemsView = this.gems.map((gem) => gem.toString()).join("\n");

        if (this.imbuedSupport !== "") {
            return `<Skill enabled="true" slot="${this.slot}" mainActiveSkill="nil" imbuedSupport="${this.imbuedSupport}">
${gemsView}
</Skill>`;
        } else {
            return `<Skill enabled="true" slot="${this.slot}" mainActiveSkill="nil">
${gemsView}
</Skill>`;
        }
    }
}

export class Gem {
    level = 20;
    qualityId = "Default";
    quality = 0;
    nameSpec = "";
    enableGlobal1 = true;
    enableGlobal2 = false;

    private constructor() {}

    static parseJson(json: itemTypes.Gem): {
        gem: Gem;
        buildInSupport: Gem | undefined;
    } {
        const gem = new Gem();
        let buildInSupport: Gem | undefined = undefined;

        const propMap = new Map<string, itemTypes.Property>();
        if (json.properties) {
            json.properties.forEach((prop) => propMap.set(prop.name, prop));
        }
        gem.level = parseIntOrDefault(propMap.get("Level")?.values[0][0], 20);
        try {
            gem.quality = parseIntOrDefault(
                propMap.get("Quality")?.values[0][0],
                0,
            );
        } catch (e) {
            console.log(json);
            throw e;
        }

        gem.nameSpec = json.baseType.replace(" Support", "");
        if (json.hybrid && json.hybrid.isVaalGem) {
            const hybridBaseTypeName = json.hybrid.baseTypeName;
            if (isTransfiguredSkill(hybridBaseTypeName)) {
                gem.nameSpec =
                    gem.nameSpecOfVaalTransfiguredGem(hybridBaseTypeName);
            }
        }

        if (gem.isVaalGem()) {
            gem.enableGlobal1 = false;
            gem.enableGlobal2 = true;
        }

        if (json.builtInSupport) {
            // 从类似`Supported by Level 1 Damage on Full Life`的文本中提取出`1`,`Damage on Full Life`
            const builtInSupport = json.builtInSupport;
            const match = builtInSupport.match(/Supported by Level (\d+) (.+)/);
            if (match) {
                const level = match[1];
                const support = match[2];
                buildInSupport = new Gem();
                buildInSupport.level = parseIntOrDefault(level, 1);
                buildInSupport.nameSpec = support;
            } else {
                console.error(
                    `Failed to parse gem builtInSupport: ${builtInSupport}`,
                );
            }
        }

        return { gem, buildInSupport };
    }

    nameSpecOfVaalTransfiguredGem(transfiguredGemName: string): string {
        return "Vaal " + transfiguredGemName;
    }

    isVaalGem() {
        return this.nameSpec.startsWith("Vaal ");
    }

    public toString(): string {
        return `<Gem level="${this.level}" qualityId="${this.qualityId}" \
quality="${this.quality}" nameSpec="${this.nameSpec}" enabled="true" \
enableGlobal1="${this.enableGlobal1}" enableGlobal2="${this.enableGlobal2}"/>`;
    }
}

File: go/building/xml/skill.go
package xml

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/cn-poe-community/cn-poe-utils/go/api"
	"github.com/cn-poe-community/cn-poe-utils/go/building/util"
)

// Skills 技能集合
type Skills struct {
	SkillSet *SkillSet
}

func NewSkills() *Skills {
	return &Skills{
		SkillSet: NewSkillSet(),
	}
}

// String 返回XML字符串
func (s *Skills) String() string {
	return fmt.Sprintf(`<Skills activeSkillSet="1">
%s
</Skills>`, s.SkillSet.String())
}

// SkillSet 技能集合
type SkillSet struct {
	Skills []*Skill
}

func NewSkillSet() *SkillSet {
	return &SkillSet{
		Skills: []*Skill{},
	}
}

// String 返回XML字符串
func (ss *SkillSet) String() string {
	var skillsView []string
	for _, skill := range ss.Skills {
		skillsView = append(skillsView, skill.String())
	}
	return fmt.Sprintf(`<SkillSet id="1">
%s
</SkillSet>`, strings.Join(skillsView, "\n"))
}

// Skill 技能
type Skill struct {
	Slot          string
	Gems          []*Gem
	ImbuedSupport string
}

func NewSkill(slotName string, jsonList []*api.Item) *Skill {
	skill := &Skill{
		Slot: slotName,
		Gems: []*Gem{},
	}
	for _, json := range jsonList {
		gem, buildInSupport := parseJson(json)
		skill.Gems = append(skill.Gems, gem)
		if buildInSupport != nil {
			skill.ImbuedSupport = buildInSupport.NameSpec
		}
	}
	return skill
}

// String 返回XML字符串
func (s *Skill) String() string {
	var gemsView []string
	for _, gem := range s.Gems {
		gemsView = append(gemsView, gem.String())
	}
	if s.ImbuedSupport == "" {
		return fmt.Sprintf(`<Skill enabled="true" slot="%s" mainActiveSkill="nil">
%s
</Skill>`, s.Slot, strings.Join(gemsView, "\n"))
	}
	return fmt.Sprintf(`<Skill enabled="true" slot="%s" mainActiveSkill="nil" imbuedSupport="%s">
%s
</Skill>`, s.Slot, s.ImbuedSupport, strings.Join(gemsView, "\n"))
}

// Gem 宝石
type Gem struct {
	Level         int
	QualityId     string
	Quality       int
	NameSpec      string
	EnableGlobal1 bool
	EnableGlobal2 bool
}

func parseJson(json *api.Item) (*Gem, *Gem) {
	propMap := make(map[string]*api.Property)
	if json.Properties != nil {
		for i := range json.Properties {
			prop := &json.Properties[i]
			propMap[prop.Name] = prop
		}
	}

	level := 20
	if propMap["Level"] != nil && len(propMap["Level"].Values) > 0 {
		level = util.ParseIntOrDefault(propMap["Level"].Values[0].Value, 20)
	}

	quality := 0
	if propMap["Quality"] != nil && len(propMap["Quality"].Values) > 0 {
		quality = util.ParseIntOrDefault(propMap["Quality"].Values[0].Value, 0)
	}

	nameSpec := strings.Replace(json.BaseType, " Support", "", 1)

	if json.Hybrid != nil && json.Hybrid.IsVaalGem != nil && *json.Hybrid.IsVaalGem {
		hybridBaseTypeName := json.Hybrid.BaseTypeName
		if util.IsTransfiguredSkill(hybridBaseTypeName) {
			nameSpec = nameSpecOfVaalTransfiguredGem(hybridBaseTypeName)
		}
	}

	gem := &Gem{
		Level:         level,
		QualityId:     "Default",
		Quality:       quality,
		NameSpec:      nameSpec,
		EnableGlobal1: true,
		EnableGlobal2: false,
	}

	if gem.IsVaalGem() {
		gem.EnableGlobal1 = false
		gem.EnableGlobal2 = true
	}

	var buildInSupport *Gem

	// 处理内置辅助技能
	if json.BuiltInSupport != nil {
		builtInSupport := *json.BuiltInSupport
		re := regexp.MustCompile(`Supported by Level (\d+) (.+)`)
		match := re.FindStringSubmatch(builtInSupport)
		if len(match) == 3 {
			supportLevel := util.ParseIntOrDefault(match[1], 1)
			supportName := match[2]
			buildInSupport = &Gem{
				Level:         supportLevel,
				QualityId:     "Default",
				Quality:       0,
				NameSpec:      supportName,
				EnableGlobal1: true,
				EnableGlobal2: false,
			}
		} else {
			fmt.Printf("Failed to parse gem builtInSupport: %s\n", builtInSupport)
		}
	}

	return gem, buildInSupport
}

func nameSpecOfVaalTransfiguredGem(transfiguredGemName string) string {
	return "Vaal " + transfiguredGemName
}

// IsVaalGem 判断是否为瓦尔宝石
func (g *Gem) IsVaalGem() bool {
	return len(g.NameSpec) > 5 && g.NameSpec[:5] == "Vaal "
}

// String 返回XML字符串
func (g *Gem) String() string {
	return fmt.Sprintf(`<Gem level="%d" qualityId="%s" quality="%d" nameSpec="%s" enabled="true" enableGlobal1="%t" enableGlobal2="%t"/>`,
		g.Level, g.QualityId, g.Quality, g.NameSpec, g.EnableGlobal1, g.EnableGlobal2)
}
